Mod2/Lessons/Week2/TDD.md

```diff
@@ -0,0 +1,152 @@
+
+<details><summary>Class Outline</summary><br/>
+![Class Ouline](/Mod2/Images/Week2/littleshopclassoutline.png)
+</details>
+
+### Writing our First Test
+
+We are going to start by creating tests for the Item class (because it is the simplest class 🙂).
+
+>Individually:
+>1. Create a new Console Application in Visual Studio, called `LittleShop`.
+>2. Add a test project called `LittleShop.Tests`.
+>3. Create the tests for a future Item class.
+
+** Instructor note ** After giving them some time individually to create tests, put them into groups to review what they came up with.
+
+>In small groups, review the tests you each created.  Take turns sharing your screens and asking questions.  Be ready to share out!
+
+<details><summary>Item Tests</summary><br/>
+![](/Mod2/Images/Week2/ItemTest.png)
+</details>
+
+With this test created, we can see that Visual Studio is already giving us some insight into what we should do first in our implementation.  We see that `Item` is underlined in red - meaning that we don't have an Item class in scope. We are going to use feedback from our tests to determine how to implement the class.
+
+### Implement through Errors
+
+Using TDD, we want to implement **only** the code required to clear a specific error or issue, then address the next issue as it arises.  So, at this point we are not going to implement the entire Item class, we are only going to implement the code needed to pull an Item class into scope.
+
+> With a partner, take 5 minutes to determine the smallest amount of coding we need to do in order to resolve this first issue.
+
+<details><summary>Solution</summary><br/>
+The only thing we need to do is create an Item class (and add the project reference).  Doing that will clear the first issue, and presents us with new errors!
+
+![](/Mod2/Images/Week2/ItemClassCreated.png)
+</details>
+
+One of the biggest benefits of TDD is that we _only create the code that we need_.  Following TDD should ensure that we are not creating unnecessary code.  But, to get this benefit, we need to make **small, incremental** changes.  You should repeat the following steps:
+
+1. Identify a single issue
+2. Make the smallest change possible to fix that issue
+3. Repeat step 1
+
+>With your partner, follow this incremental process to complete the Item class.  Keep track of how many errors/issues you come across and be ready to share out!
+
+** Instructor Note ** Be sure to ask them at what point they started running the test to see what to do next.  It is possible (likely) that students have not run the tests at all - they may rely only on the text editor's feedback to determine what to do next. Following the steps above, they _should_ get to a point where the properties are created but are not assigned to the constructor argument values.
+
+<details><summary>Completed Item Class</summary><br/>
+![](/Mod2/Images/Week2/ItemClassComplete.png)
+</details>
+
+
+### TDD the Customer and Order Class
```

Comment:
Making a note to capture what we talked about in person:

Going to simplify the customer class to only have name and address.
Going to add an overarching store class to allow for integration testing.

Going to stick with writing all tests, then writing all code to pass the tests. But planning to add a bit of discussion around how you might do TDD in the real world. Write failing test, write code to pass test, refactor, repeat.